### ISSUE TITLE
Mega Android SDK: Cancel streaming download crashs JNI and app

### ISSUE THREAD

**Issue by cocia** [meganz/sdk#253]
I am using the latest code in master(2016-03-30) to build libmega.so for x86 and arm.
I tried to use stream download in MegaApiAndroid to download file from specific offset of file in Android 4.x.x. Stream download works fine by api:
public void startStreaming(MegaNode node, long startPos, long size, MegaTransferListenerInterface listener) 

But app will crash when I cancel the stream download by api:
public void cancelTransfer(MegaTransfer transfer, MegaRequestListenerInterface listener)
The transfer object is got from MegaTransferListenerInterface callback.

I tried when I cancel the download which is created by method startDownload(non stream download method), it works fine. 

This issue happens in both x86 and arm devices and print similar log. 

What's wrong in my usage or in sdk?
Thanks.

============================error log begin=================================
04-01 15:45:21.778        86-86/? I/DEBUG﹕ *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
04-01 15:45:21.778        86-86/? I/DEBUG﹕ Build fingerprint: 'generic/vbox86tp/vbox86tp:5.1/LMY47D/buildbot11172044:userdebug/test-keys'
04-01 15:45:21.778        86-86/? I/DEBUG﹕ Revision: '0'
04-01 15:45:21.778        86-86/? I/DEBUG﹕ ABI: 'x86'
04-01 15:45:21.778        86-86/? I/DEBUG﹕ pid: 3540, tid: 3564, name: Thread-1576  >>> com.example.cloudapi_sample.app <<<
04-01 15:45:21.778        86-86/? I/DEBUG﹕ signal 11 (SIGSEGV), code 1 (SEGV_MAPERR), fault addr 0x30
04-01 15:45:21.782        86-86/? I/DEBUG﹕ eax 00000030  ebx e2d2bcc8  ecx e2d2bcc8  edx e2d20010
04-01 15:45:21.782        86-86/? I/DEBUG﹕ esi 00000000  edi e2938ff0
04-01 15:45:21.782        86-86/? I/DEBUG﹕ xcs 00000023  xds 0000002b  xes 0000002b  xfs 00000000  xss 0000002b
04-01 15:45:21.782        86-86/? I/DEBUG﹕ eip e29cd278  ebp e2939cf8  esp e2938f50  flags 00210282
04-01 15:45:21.783        86-86/? I/DEBUG﹕ backtrace:
04-01 15:45:21.783        86-86/? I/DEBUG﹕ #00 pc 00093278  /data/app/com.example.cloudapi_sample.app-1/lib/x86/libmega.so
04-01 15:45:21.783        86-86/? I/DEBUG﹕ #01 pc 00097447  /data/app/com.example.cloudapi_sample.app-1/lib/x86/libmega.so
04-01 15:45:21.783        86-86/? I/DEBUG﹕ #02 pc 000974f7  /data/app/com.example.cloudapi_sample.app-1/lib/x86/libmega.so
04-01 15:45:21.783        86-86/? I/DEBUG﹕ #03 pc 000211d8  /system/lib/libc.so (__pthread_start(void*)+56)
04-01 15:45:21.783        86-86/? I/DEBUG﹕ #04 pc 0001c559  /system/lib/libc.so (__start_thread+25)
04-01 15:45:21.783        86-86/? I/DEBUG﹕ #05 pc 00013126  /system/lib/libc.so (__bionic_clone+70)
04-01 15:45:22.021        86-86/? I/DEBUG﹕ Tombstone written to: /data/tombstones/tombstone_06
04-01 15:45:22.022    1929-1951/system_process I/BootReceiver﹕ Copying /data/tombstones/tombstone_06 to DropBox (SYSTEM_TOMBSTONE)
============================error log end=====================================

**Comment by javiserrano** [comment]
Your usage seems to be OK.
We will try to reproduce the problem and fix it.

In order to reproduce it easier, could you answer these questions?
- Does the app crashes when you use `cancelTransfer` with a transfer started with `startDownload`, `startStreaming`, or both? At the beginning of your report you mention `startStreaming`, but a bit later you talk about `startDownload`.
- Are you using `MegaApiAndroid` or `MegaApi`? Please use `MegaApiAndroid`, otherwise the parameters of callbacks aren't valid after callbacks finish. Anyway, since you say that you have got the `MegaTransfer` object using `MegaTransferListenerInterface`, I suppose that you are already using `MegaApiAndroid`.

**Comment by cocia** [comment]
I have corrected the original issue text by your question and summarize here:
startDownload and cancel, it worked.
startStreaming and cancel,it crashed.
Mega API is "MegaApiAndroid".

Actually first I tried SDK(2015-5-5) with same test result which was got from Mega net disk(https://github.com/meganz/sdk/tree/master/examples/android). So I upgraded to latest version and tried again but not fortunate.

**Comment by javiserrano** [comment]
Thank you.

I have reproduced the problem. The cause was that `cancelTransfer` wasn't prepared to cancel streaming transfers. All of our apps currently return `false` on `onTransferData` to cancel them.

This problem and some others related to streaming transfers were already fixed in a different branch (`jssdk`). Since that branch contains a lot of changes and could take a long time to be properly reviewed, tested and merged, I have extracted the changes related to streaming transfers:
- New function `getStreamingTransfers` in `MegaApi`
- Compatibility with streaming transfers in `cancelTransfer`
- Provide the transfer speed on `MegaTransfer::getSpeed` on callbacks related to streaming transfers

The current code in `master` should allow you to use `cancelTransfer` without crashes.
